File: programming/task-7/Validation.py
import re
from datetime import date    

class Validation:
    def __init__(self): 
        pass

    @staticmethod
    def validateFloat(func):
        def inner (_self, val):
            try:
                val = round(float(val), 2)
                func(_self, val)
            except ValueError: 
                raise ValueError('Value must be float type')            

        return inner

    @staticmethod
    def validateFloatInRange(func):
        def inner (_self, val):
            MIN_VAL, MAX_VAL = 0, 5000
            if not (MIN_VAL < val and val <= MAX_VAL): 
                raise ValueError('Value must be less than %d' %(MAX_VAL))   
            func(_self, val)        

        return inner      

    @staticmethod
    def validateStr(func):
        def inner (_self, val):
            if not isinstance(val, str):
                raise ValueError('Value must be str type')
            func(_self, val)            

        return inner     

    @staticmethod
    def validateTitle(func):
        def inner (_self, val):
            if re.search(r'[^A-Za-z ]+', val) is not None: 
                raise ValueError('Value must contain only letters')    
            func(_self, val)            

        return inner

    @staticmethod
    def validateURL(func):
        def inner (_self, val):
            regexSrch = re.search(r'^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&\'\(\)\*\+,;=.]+$', val)
            if regexSrch is None: 
                raise ValueError('Bad URL value')            
            func(_self, val)

        return inner

    @staticmethod
    def validateDate(func):
        def inner (_self, val):
            try:
                if isinstance(val, str):
                    [yyyy, mm, dd] = val.split('-')
                    val = date(int(yyyy), int(mm), int(dd))
            except ValueError:
                raise ValueError('Value must be Date type')
            func(_self, val)

        return inner

    @staticmethod
    def validateTwoDates(func):
        def inner(_self, val1, val2):
            if val1 > val2:
                raise ValueError('Bad Date values')
            func(_self, val1, val2)

        return inner

    @staticmethod
    def validateFileName(func):
        def inner (_self, val): 
            if  not (val.endswith('.json') or val.endswith('.txt')):
                raise ValueError('Value must be json or txt format')
            func(_self, val)

        return inner

    @staticmethod
    def validateProduct(func):
        def inner (_self, val):
            Validation.validateStr(func)
            Validation.validateFloatInRange(func)
            Validation.validateStr(func)
            Validation.validateURL(func)
            Validation.validateDate(func)
            Validation.validateDate(func)
            func(_self, val)
            return val

        return inner

File: csharp/edu-practice/edu-practice-2/edu-practice-2/Validation.cs
using System;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;

namespace edu_practice_2
{
    public static class Validation
    {
        public static string ValidateName(string value)
        {
            if (value.All(c => Char.IsLetter(c) || c == ' '))
            {
                return value;
            }
            throw new FormatException($"String value must be a string containing only letters and spaces, {value} was given");
        }

        public static string ValidateUrl(string value)
        {
            string pattern = @"[(http(s)?):\/\/(www\.)?a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)";
            if (Regex.IsMatch(value, pattern))
            {
                return value;
            }
            throw new FormatException($"String value must be a URL, {value} was given");
        }

        public static double ValidatePrice(double value)
        {
            if (value > 0.0)
            {
                return Math.Round(value, 2);
            }

            throw new ArgumentOutOfRangeException("Price", $"Float value must be positive, {value} was given");
        }

        public static DateTime ValidateDate(DateTime value, DateTime other = new DateTime())
        {
            if (value < DateTime.Now && (other == DateTime.MinValue || value >= other))
            {
                return value;
            }
            throw new ArgumentOutOfRangeException("UpdatedAt", $"DateTime value must be from past (less or equal than CreatedAt), {value} was given");
        }
        
        public static string ValidateFileName(string value)
        {
            if (value.EndsWith(".json") || value.EndsWith(".txt"))
            {
                return value;
            }
            throw new FormatException($"String value must be .json ot .txt format, {value} was given");
        }

        public static string ValidateProperty(string value, Type type)
        {
            string[] props = type.GetProperties(BindingFlags.Instance | BindingFlags.Public)
                .Select(p => p.Name).ToArray();

            if (props.Contains(value))
            {
                return value;
            }

            throw new MissingFieldException($"{type.Name} has no property {value} ");
        }
    }
}